tarantool/connection_pool.py

```diff
@@ -0,0 +1,466 @@
+# -*- coding: utf-8 -*-
+
+import abc
+import itertools
+import queue
+import threading
+import time
+import typing
+from dataclasses import dataclass, field
+from enum import Enum
+
+from tarantool.connection import Connection, ConnectionInterface
+from tarantool.const import (
+    CONNECTION_TIMEOUT,
+    POOL_INSTANCE_RECONNECT_DELAY,
+    POOL_INSTANCE_RECONNECT_MAX_ATTEMPTS,
+    POOL_REFRESH_DELAY,
+    SOCKET_TIMEOUT
+)
+from tarantool.error import (
+    ClusterConnectWarning,
+    PoolTolopogyError,
+    PoolTolopogyWarning,
+    ConfigurationError,
+    DatabaseError,
+    NetworkError,
+    NetworkWarning,
+    tnt_strerror,
+    warn
+)
+from tarantool.utils import ENCODING_DEFAULT
+from tarantool.mesh_connection import validate_address
+
+
+class Mode(Enum):
+    ANY = 1
+    RW = 2
+    RO = 3
+    PREFER_RW = 4
+    PREFER_RO = 5
+
+
+class Status(Enum):
+    HEALTHY = 1
+    UNHEALTHY = 2
+
+
+@dataclass
+class InstanceState():
+    status: Status = Status.UNHEALTHY
+    ro: typing.Optional[bool] = None
+
+
+def QueueFactory():
+    return queue.Queue(maxsize=1)
+
+
+@dataclass
+class PoolUnit():
+    addr: dict
+    conn: Connection
+    input_queue: queue.Queue = field(default_factory=QueueFactory)
+    output_queue: queue.Queue = field(default_factory=QueueFactory)
+    thread: typing.Optional[threading.Thread] = None
+    state: InstanceState = field(default_factory=InstanceState)
+    # request_processing_enabled is used to stop requests processing
+    # in background thread on close or destruction.
+    request_processing_enabled: bool = False
+
+
+# Based on https://realpython.com/python-interface/
+class StrategyInterface(metaclass=abc.ABCMeta):
+    @classmethod
+    def __subclasshook__(cls, subclass):
+        return (hasattr(subclass, 'update') and
+                callable(subclass.update) and
+                hasattr(subclass, 'getnext') and
+                callable(subclass.getnext) or
+                NotImplemented)
+
+    @abc.abstractmethod
+    def update(self):
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def getnext(self, mode):
+        raise NotImplementedError
+
+class RoundRobinStrategy(object):
+    """
+    Simple round-robin connection rotation
+    """
+    def __init__(self, pool):
+        self.ANY_iter = None
+        self.RW_iter = None
+        self.RO_iter = None
+        self.pool = pool
+        self.rebuild_needed = True
+
+    def build(self):
+        ANY_pool = []
+        RW_pool = []
+        RO_pool = []
+
+        for key in self.pool:
+            state = self.pool[key].state
+
+            if state.status == Status.UNHEALTHY:
+                continue
+
+            ANY_pool.append(key)
```

Comment:
It seems logical, but I have a question: maybe it makes sense to have only ro/rw pools and choose a connection from the pool where there are more connections. So the load would be distributed evenly but on the other hand we will have a separate patch with balancer. What do you think? I'm not sure if any changes are needed, I just wanted to share my thoughts.

Reply:
> <...> choose a connection from the pool where there are more connections. So the load would be distributed evenly <...>

I disagree here. Say, we have one master and two replicas. All requests with the `ANY` mode will go to the replicas (we have more RO instances than RW ones). It is not uniform distribution.

I think that it is technically possible to eliminate the `ANY_pool` list, but it'll add extra code complexity.

Reply:
I agree with Totktonada